File: src/main/java/info/skyblond/ariteg/storage/disk/SerializerByteString.java
package info.skyblond.ariteg.storage.disk;

import com.google.protobuf.ByteString;
import org.jetbrains.annotations.NotNull;
import org.mapdb.DataInput2;
import org.mapdb.DataOutput2;
import org.mapdb.serializer.GroupSerializer;
import org.mapdb.serializer.SerializerByteArray;

import java.io.IOException;
import java.util.Comparator;

public class SerializerByteString implements GroupSerializer<ByteString> {
    private final SerializerByteArray serializer = new SerializerByteArray();

    @Override
    public int valueArraySearch(Object keys, ByteString key) {
        return this.serializer.valueArraySearch(keys, key.toByteArray());
    }

    @Override
    public int valueArraySearch(Object keys, ByteString key, Comparator comparator) {
        return this.serializer.valueArraySearch(keys, key.toByteArray(), comparator);
    }

    @Override
    public void valueArraySerialize(DataOutput2 out, Object vals) throws IOException {
        this.serializer.valueArraySerialize(out, vals);
    }

    @Override
    public Object valueArrayDeserialize(DataInput2 in, int size) throws IOException {
        return this.serializer.valueArrayDeserialize(in, size);
    }

    @Override
    public ByteString valueArrayGet(Object vals, int pos) {
        return ByteString.copyFrom(this.serializer.valueArrayGet(vals, pos));
    }

    @Override
    public int valueArraySize(Object vals) {
        return this.serializer.valueArraySize(vals);
    }

    @Override
    public Object valueArrayEmpty() {
        return this.serializer.valueArrayEmpty();
    }

    @Override
    public Object valueArrayPut(Object vals, int pos, ByteString newValue) {
        return this.serializer.valueArrayPut(vals, pos, newValue.toByteArray());
    }

    @Override
    public Object valueArrayUpdateVal(Object vals, int pos, ByteString newValue) {
        return this.serializer.valueArrayUpdateVal(vals, pos, newValue.toByteArray());
    }

    @Override
    public Object valueArrayFromArray(Object[] objects) {
        return this.serializer.valueArrayFromArray(objects);
    }

    @Override
    public Object valueArrayCopyOfRange(Object vals, int from, int to) {
        return this.serializer.valueArrayCopyOfRange(vals, from, to);
    }

    @Override
    public Object valueArrayDeleteValue(Object vals, int pos) {
        return this.serializer.valueArrayDeleteValue(vals, pos);
    }

    @Override
    public void serialize(@NotNull DataOutput2 out, @NotNull ByteString value) throws IOException {
        this.serializer.serialize(out, value.toByteArray());
    }

    @Override
    public ByteString deserialize(@NotNull DataInput2 input, int available) throws IOException {
        return ByteString.copyFrom(this.serializer.deserialize(input, available));
    }
}

File: src/main/kotlin/info/skyblond/ariteg/storage/client/disk/SerializerByteString.kt
package info.skyblond.ariteg.storage.client.disk

import org.mapdb.serializer.GroupSerializer
import com.google.protobuf.ByteString
import org.mapdb.serializer.SerializerByteArray
import kotlin.Throws
import java.io.IOException
import org.mapdb.DataOutput2
import org.mapdb.DataInput2
import java.util.Comparator

class SerializerByteString : GroupSerializer<ByteString> {
    private val serializer = SerializerByteArray()
    override fun valueArraySearch(keys: Any, key: ByteString): Int {
        return serializer.valueArraySearch(keys, key.toByteArray())
    }

    override fun valueArraySearch(keys: Any, key: ByteString, comparator: Comparator<*>?): Int {
        return serializer.valueArraySearch(keys, key.toByteArray(), comparator)
    }

    @Throws(IOException::class)
    override fun valueArraySerialize(out: DataOutput2, vals: Any) {
        serializer.valueArraySerialize(out, vals)
    }

    @Throws(IOException::class)
    override fun valueArrayDeserialize(`in`: DataInput2, size: Int): Any {
        return serializer.valueArrayDeserialize(`in`, size)
    }

    override fun valueArrayGet(vals: Any, pos: Int): ByteString {
        return ByteString.copyFrom(serializer.valueArrayGet(vals, pos))
    }

    override fun valueArraySize(vals: Any): Int {
        return serializer.valueArraySize(vals)
    }

    override fun valueArrayEmpty(): Any {
        return serializer.valueArrayEmpty()
    }

    override fun valueArrayPut(vals: Any, pos: Int, newValue: ByteString): Any {
        return serializer.valueArrayPut(vals, pos, newValue.toByteArray())
    }

    override fun valueArrayUpdateVal(vals: Any, pos: Int, newValue: ByteString): Any {
        return serializer.valueArrayUpdateVal(vals, pos, newValue.toByteArray())
    }

    override fun valueArrayFromArray(objects: Array<Any>): Any {
        return serializer.valueArrayFromArray(objects)
    }

    override fun valueArrayCopyOfRange(vals: Any, from: Int, to: Int): Any {
        return serializer.valueArrayCopyOfRange(vals, from, to)
    }

    override fun valueArrayDeleteValue(vals: Any, pos: Int): Any {
        return serializer.valueArrayDeleteValue(vals, pos)
    }

    @Throws(IOException::class)
    override fun serialize(out: DataOutput2, value: ByteString) {
        serializer.serialize(out, value.toByteArray())
    }

    @Throws(IOException::class)
    override fun deserialize(input: DataInput2, available: Int): ByteString {
        return ByteString.copyFrom(serializer.deserialize(input, available))
    }
}
